src/Pact/Analyze/Types/Model.hs

```diff
@@ -27,49 +31,114 @@ data Arg = Arg
 
 newtype TagId
   = TagId Natural
-  deriving (Num, Show, Ord, Eq)
+  deriving (Num, Enum, Show, Ord, Eq)
 
-data TagAllocation
-  = AllocReadTag (Located (TagId, Schema))
-  | AllocWriteTag (Located (TagId, Schema))
-  | AllocAuthTag (Located TagId)
-  | AllocVarTag (Located (VarId, Text, EType))
-  deriving Show
+type Edge = (Vertex, Vertex)
 
-data ModelTags
+newtype Vertex
+  = Vertex Natural
+  deriving (Num, Enum, Show, Ord, Eq)
+
+data Recoverability
+  -- The path upon which to resume inclusion of events. An alternative here
+  -- would be to have more edges in the subgraphs formed by @enforce-one@ --
+  -- have vertices not just for each case, but additionally one after each
+  -- recoverable assert/auth as well, to connect from right after the
+  -- assert/auth to the next case. One would need to be careful here not to try
+  -- to make two edges between a pair of vertices when an assert/auth in a
+  -- preceeding case is right before the next case. We need to explicitly talk
+  -- about the path upon which to resume execution because it's not necessarily
+  -- the next vertex -- there could be more subpaths before where we should
+  -- resume due to nested conditionals or @enforce-one@s.
+  = Recoverable { _resumptionPath :: TagId }
+  | Unrecoverable
+  deriving (Eq, Show)
+
+-- For determining resumption paths for nested @enforce-one@s:
+instance Semigroup Recoverability where
+  --      outer <> inner
+  Recoverable _ <> Recoverable y = Recoverable y
+  Recoverable x <> Unrecoverable = Recoverable x
+  Unrecoverable <> nested        = nested
+
+instance Monoid Recoverability where
+  mempty = Unrecoverable
+  mappend = (<>)
+
+data TraceEvent
+  = TraceRead (Located (TagId, Schema))
+  | TraceWrite (Located (TagId, Schema))
+  | TraceAssert Recoverability (Located TagId)
+  | TraceAuth Recoverability (Located TagId)
+  | TraceBind (Located (VarId, Text, EType))
+  | TraceSubpathStart TagId
+  deriving (Eq, Show)
+
+data ExecutionGraph
```

Comment:
I'd love notes about `ExecutionGraph` / `ExecutionTrace`. What are they, how do they fit together?

Reply:
Is `ExecutionTrace` even used?

Reply:
I can add notes for them. `ExecutionTrace` is indeed used -- it is produced by `linearizedTrace :: Model 'Concrete -> ExecutionTrace` for presentation.